### ISSUE TITLE
[Documentation] Add Migration Guide from PyTorch

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2675]
## Problem

PyTorch users need guidance on converting models and workflows to ML Odyssey.

## Proposed: docs/migration/pytorch-to-ml-odyssey.md

```markdown
# PyTorch to ML Odyssey Migration Guide

Guide for migrating PyTorch models and code to ML Odyssey.

## Quick Comparison

| Concept | PyTorch | ML Odyssey |
|---------|---------|------------|
| **Tensor** | `torch.Tensor` | `ExTensor` |
| **Module** | `nn.Module` | `Module` |
| **Autograd** | Automatic | `Tape` |
| **Optimizer** | `torch.optim.SGD` | `SGD` |
| **Device** | `.to('cuda')` | (CPU only for now) |

## Tensor Operations

### Creation

```python
# PyTorch
import torch
x = torch.tensor([1, 2, 3])
y = torch.zeros(10, 10)
z = torch.randn(5, 5)
```

```mojo
# ML Odyssey
from shared.core import ExTensor
var x = ExTensor([1, 2, 3], DType.float32)
var y = zeros([10, 10], DType.float32)
var z = random_tensor([5, 5], DType.float32)
```

### Basic Operations

```python
# PyTorch
c = a + b
d = torch.matmul(a, b)
e = a.reshape(2, 3)
```

```mojo
# ML Odyssey
var c = a + b
var d = matmul(a, b)
var e = a.reshape([2, 3])
```

### Indexing

```python
# PyTorch
x[0]
x[:, 1]
x[0:5, 2:7]
```

```mojo
# ML Odyssey
x[0]
x.slice(axis=1, start=1, end=2)  # Column 1
x.slice_2d(0, 5, 2, 7)  # Rows 0-5, cols 2-7
```

### PyTorch Model

```python
import torch.nn as nn

class SimpleModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(784, 128)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(128, 10)
    
    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        return x
```

### ML Odyssey Model

```mojo
from shared.nn import Module, Linear, ReLU

struct SimpleModel(Module):
    var fc1: Linear
    var relu: ReLU
    var fc2: Linear
    
    fn __init__(out self):
        self.fc1 = Linear(784, 128)
        self.relu = ReLU()
        self.fc2 = Linear(128, 10)
    
    fn forward(self, x: ExTensor) -> ExTensor:
        var out = self.fc1.forward(x)
        out = self.relu.forward(out)
        out = self.fc2.forward(out)
        return out
    
    fn parameters(self) -> List[ExTensor]:
        var params = List[ExTensor]()
        params.extend(self.fc1.parameters())
        params.extend(self.fc2.parameters())
        return params
```

### PyTorch

```python
model = SimpleModel()
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
criterion = nn.CrossEntropyLoss()

for epoch in range(10):
    for batch in dataloader:
        optimizer.zero_grad()
        
        output = model(batch['input'])
        loss = criterion(output, batch['target'])
        
        loss.backward()
        optimizer.step()
```

### ML Odyssey

```mojo
var model = SimpleModel()
var optimizer = SGD(model.parameters(), lr=0.01)
var criterion = CrossEntropyLoss()
var tape = Tape()

for epoch in range(10):
    for batch in dataloader:
        optimizer.zero_grad()
        
        var output = model.forward(batch.input)
        var loss = criterion.forward(output, batch.target)
        
        tape.backward(loss)
        optimizer.step()
```

### 1. Autograd

**PyTorch**: Automatic gradient tracking

```python
x = torch.tensor([1.0], requires_grad=True)
y = x * 2
y.backward()
print(x.grad)  # Gradient automatically computed
```

**ML Odyssey**: Explicit tape

```mojo
var tape = Tape()
var x = ExTensor([1.0], DType.float32, requires_grad=True)
var y = x * 2
tape.backward(y)
print(x.grad)  # Gradient computed via tape
```

### 2. Module Parameters

**PyTorch**: Automatic parameter registration

```python
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(10, 5)  # Parameters auto-registered
```

**ML Odyssey**: Manual parameter collection

```mojo
struct Model(Module):
    var fc: Linear
    
    fn parameters(self) -> List[ExTensor]:
        return self.fc.parameters()  # Must collect manually
```

### 3. In-Place Operations

**PyTorch**: Common, denoted with `_`

```python
x.add_(1)  # In-place add
x.relu_()  # In-place ReLU
```

**ML Odyssey**: Explicit mutability

```mojo
x.add_inplace(1)  # Must have mut self
x = relu(x)  # Usually return new tensor
```

## Model Conversion Workflow

1. **Define model structure** in Mojo
2. **Train in PyTorch** (for now, until all ops implemented)
3. **Export PyTorch weights** to numpy
4. **Load weights** into ML Odyssey model
5. **Verify outputs** match

### Example: Converting Weights

```python
# PyTorch: Save weights
torch.save(model.state_dict(), 'weights.pth')

# Convert to numpy
import numpy as np
weights = {}
for name, param in model.state_dict().items():
    weights[name] = param.cpu().numpy()
np.savez('weights.npz', **weights)
```

```mojo
# ML Odyssey: Load weights
var weights = load_npz("weights.npz")
model.load_state_dict(weights)
```

### 1. Shape Conventions

PyTorch uses channels-first (NCHW), ML Odyssey same:
- `(batch, channels, height, width)` ✓

### 2. Default DTypes

```python
# PyTorch default: float32
x = torch.tensor([1.0])  # float32
```

```mojo
# ML Odyssey: Must specify
var x = ExTensor([1.0], DType.float32)  # Explicit
```

### 3. Gradient Accumulation

```python
# PyTorch: Must call zero_grad()
optimizer.zero_grad()
```

```mojo
# ML Odyssey: Same
optimizer.zero_grad()
```

## What's Missing

Currently ML Odyssey doesn't have:
- GPU support (coming soon)
- Some advanced layers (Transformer, etc.)
- Distributed training
- torch.jit equivalent

See [Roadmap](roadmap.md) for planned features.

## Resources

- [PyTorch Documentation](https://pytorch.org/docs/)
- [ML Odyssey API Reference](../api/)
- [Examples](../../examples/)
```

## Benefits

- Help PyTorch users transition
- Highlight API similarities/differences
- Reduce learning curve
- Build community

## Impact

- **Severity**: LOW
- **Adoption**: Easier for PyTorch users
- **Documentation**: Complete migration story

**Comment by mvillmow** [comment]
❌ Implementation failed:

```
Implementation failed after 3 attempts: Exit code 1: You've hit your limit · resets 4pm (America/Los_Angeles)

```

---
*Automated by implement_issues.py at 2025-12-19 13:18:25*
